src/runtime/tests/refiner-test.ts

```diff
@@ -17,338 +17,362 @@ import {Schema} from '../schema.js';
-        let manifestAst = parse(`
-            particle Foo
-                input: reads Something {num: Number [ ((num < 3) and (num > 0)) or (num == 5) ] }
-        `);
-        const typeData = {'num': 'Number'};
-        let ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
-        let range = Range.fromExpression(ref.expression);
-        assert.isTrue(range.equals(new Range([Segment.openOpen(0, 3), Segment.closedClosed(5, 5)])));
-        manifestAst = parse(`
-            particle Foo
-                input: reads Something {num: Number [(num > 10) == (num >= 20)] }
-        `);
-        ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
-        range = Range.fromExpression(ref.expression);
-        assert.isTrue(range.equals(new Range([Segment.openClosed(Number.NEGATIVE_INFINITY, 10), Segment.closedOpen(20, Number.POSITIVE_INFINITY)])));
-        manifestAst = parse(`
-            particle Foo
-                input: reads Something {num: Number [not (num != 10)] }
-        `);
-        ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
-        range = Range.fromExpression(ref.expression);
-        assert.isTrue(range.equals(new Range([Segment.closedClosed(10, 10)])));
-
-    }));
-    it('regression test for parse failure on operator ordering.', () => {
-      parse(`
+  it('Throws error when expression does not produce boolean result.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ num + 5 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    const ref = Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    const data = {
+      num: 6,
+    };
+    assert.throws(() => {
+      ref.validateData(data);
+    }, `Refinement expression (num + 5) evaluated to a non-boolean type.`);
+  }));
+  it('Throws error when operators and operands are incompatible.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ (num < 5) + 3 == 0 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    assert.throws(() => {
+      Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    }, `Refinement expression (num < 5) has type Boolean. Expected Number.`);
+  }));
+  it('Throws error when operators and operands are incompatible.', Flags.withFieldRefinementsAllowed(async () => {
+    const manifestAst = parse(`
+      particle Foo
+        input: reads Something {num: Number [ (num and 3) == 0 ] }
+    `);
+    const typeData = {'num': 'Number'};
+    assert.throws(() => {
+      Refinement.fromAst(manifestAst[0].args[0].type.fields[0].type.refinement, typeData);
+    }, `Refinement expression num has type Number. Expected Boolean.`);
```

Comment:
Should `num` be wrapped in single quotes? It seems like we are inconsistent with using quotes in our error messaging.

Reply:
Feel free to disregard; It might be worthwhile to do a project-wide error msg grooming post-march.

Reply:
Yes, I've got a todo to write an error handling doc if no one has does it by then.

Reply:
I'm happy to do that :)